clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,366 @@
+	klog.Infof("%s: gathering service created latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	serviceCreationLatency := s.creationTimes.CalculateTransitionsLatency(map[string]measurementutil.Transition{
+		"create_to_assigned": {
+			From: creatingPhase,
+			To:   ipAssigningPhase,
+		},
+		"assigned_to_available": {
+			From: ipAssigningPhase,
+			To:   reachabilityPhase,
+		},
+		"create_to_available": {
+			From: creatingPhase,
+			To:   reachabilityPhase,
+		},
+	})
+
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(serviceCreationLatency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, err
+}
+
+func (s *serviceCreationLatencyMeasurement) printLatencies(latencies []measurementutil.LatencyData, header string) {
+	metrics := measurementutil.NewLatencyMetric(latencies)
+	index := len(latencies) - 100
+	if index < 0 {
+		index = 0
+	}
+	klog.Infof("%s: %d %s: %v", s, len(latencies)-index, header, latencies[index:])
+	klog.Infof("%s: perc50: %v, perc90: %v, perc99: %v", s, metrics.Perc50, metrics.Perc90, metrics.Perc99)
+}
+
+type serviceCreationLatencyData struct {
+	Key     string
+	Latency time.Duration
+}
+
+func (s serviceCreationLatencyData) GetLatency() time.Duration {
+	return s.Latency
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		// Skip updates without changes in the spec.
+		return
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObjectLocked(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObjectLocked(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if pc, exists := s.pingCheckers[key]; exists {
+		pc.stop()
+		delete(s.pingCheckers, key)
+	}
+	s.creationTimes.DeleteKey(key)
```

Comment:
I don't understand it - I don't think we should ever remove anything from statistics.

Reply:
Done.